[Flang] Add a HLFIR Minloc intrinsic

davemgreen opened the pull request:
The adds a hlfir minloc intrinsic, similar to the minval intrinsic already added, to help in the lowering of minloc. The idea is to later add maxloc too, and from there add a simplification for producing minloc with inlined elemental and hopefully less temporaries.

kiranchandramohan commented:
Is the SimplifyIntrinsics pass not working anymore for Minloc? (https://github.com/llvm/llvm-project/blob/383e35048e16c85ab26bc46822d3ceb11fd4d3f2/flang/lib/Optimizer/Transforms/SimplifyIntrinsics.cpp#L1370)

Eitherway, simplifying while lowering from HLFIR Ops is probably the right direction.

jeanPerier commented on flang/lib/Lower/HlfirIntrinsics.cpp:
```diff
@@ -206,6 +219,8 @@ llvm::SmallVector<mlir::Value> HlfirTransformationalIntrinsic::getOperandVector(
       else if (!argRules.handleDynamicOptional &&
                argRules.lowerAs != fir::LowerIntrinsicArgAs::Inquired)
         valArg = hlfir::derefPointersAndAllocatables(loc, builder, actual);
+      else if (argRules.lowerAs == fir::LowerIntrinsicArgAs::Value)
```
Optional value need to use something like genOptionalValue from ConvertCall.cpp, assuming that the fallback value is zero/false.

davemgreen replied:
Ah, that's useful. Thanks. I've added a new method for conditionally loading the value. Let me know if it seems OK.

jeanPerier replied:
Looks good, thanks.

jeanPerier reviewed (commented):
Thank you for working on this

davemgreen committed:
[Flang] Add a HLFIR Minloc intrinsic

The adds a hlfir minloc intrinsic, similar to the minval intrinsic already
added, to help in the lowering of minloc. The idea is to later add maxloc too,
and from there add a simplification for producing minloc with inlined elemental
and hopefully less temporaries.

davemgreen commented:
Thanks for the review.

davemgreen commented on flang/lib/Optimizer/HLFIR/IR/HLFIROps.cpp:
```diff
@@ -870,6 +870,81 @@ void hlfir::MinvalOp::getEffects(
+  mlir::Value dim = getDim();
+  mlir::Value mask = getMask();
+
+  fir::SequenceType arrayTy =
+      hlfir::getFortranElementOrSequenceType(array.getType())
+          .cast<fir::SequenceType>();
+  llvm::ArrayRef<int64_t> arrayShape = arrayTy.getShape();
+
+  if (mask) {
+    fir::SequenceType maskSeq =
+        hlfir::getFortranElementOrSequenceType(mask.getType())
+            .dyn_cast<fir::SequenceType>();
+    llvm::ArrayRef<int64_t> maskShape;
+
+    if (maskSeq)
+      maskShape = maskSeq.getShape();
+
+    if (!maskShape.empty()) {
+      if (maskShape.size() != arrayShape.size())
+        return emitWarning("MASK must be conformable to ARRAY");
+      static_assert(fir::SequenceType::getUnknownExtent() ==
+                    hlfir::ExprType::getUnknownExtent());
+      constexpr int64_t unknownExtent = fir::SequenceType::getUnknownExtent();
+      for (std::size_t i = 0; i < arrayShape.size(); ++i) {
+        int64_t arrayExtent = arrayShape[i];
+        int64_t maskExtent = maskShape[i];
+        if ((arrayExtent != maskExtent) && (arrayExtent != unknownExtent) &&
+            (maskExtent != unknownExtent))
+          return emitWarning("MASK must be conformable to ARRAY");
+      }
+    }
+  }
```
It originally came from there, but is a bit different due to the result type not matching the input and often having a different shape than the reductions would.

jeanPerier replied:
OK, it looks the mask verification could still be shared into some verifyReductionMask helper. Custom result verification makes sense.

davemgreen commented on flang/include/flang/Optimizer/HLFIR/HLFIROps.td:
```diff
@@ -458,6 +458,33 @@ def hlfir_MinvalOp : hlfir_Op<"minval", [AttrSizedOperandSegments,
+    AnyFortranArrayObject:$array,
+    Optional<AnyIntegerType>:$dim,
+    Optional<AnyFortranLogicalOrI1ArrayObject>:$mask,
+    Optional<AnyIntegerType>:$kind,
```
I had considered that, but only went the other way as I noticed CountOp had a kind parameter already. I've had a go at changing it over to just use the return type.

jeanPerier replied:
Good point, I think we should remove it from CountOp afterwards then.

tblah reviewed (approved):
This looks great to me! Thanks for your work.

Please wait for Jean's approval as well.

jeanPerier reviewed (approved):
Looks great, thanks for addressing the comments!

davemgreen committed:
Fixup: Share more code between verify*ReductionOp

jeanPerier reviewed (approved):
LGTM
